iothub/device/src/Utilities/ObjectConversionHelper.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Globalization;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    internal class ObjectConversionHelper
+    {
+        internal static bool TryCastOrConvert<T>(object objectToCastOrConvert, PayloadConvention payloadConvention, out T value)
+        {
+            // If the object is of type T or can be cast to type T, go ahead and return it.
+            if (TryCast(objectToCastOrConvert, out value))
+            {
+                return true;
+            }
+
+            try
+            {
+                // If the cannot be cast to <T> directly we need to try to convert it using the serializer.
+                // If it can be successfully converted, go ahead and return it.
+                value = payloadConvention.PayloadSerializer.ConvertFromJsonObject<T>(objectToCastOrConvert);
+                return true;
+            }
+            catch
+            {
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal static bool TryCast<T>(object objectToCast, out T value)
+        {
+            if (objectToCast is T valueRef
+                || TryCastNumericTo(objectToCast, out valueRef))
+            {
+                value = valueRef;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal static bool TryCastNumericTo<T>(object input, out T result)
```

Comment:
private?

Reply:
this is being called by another class in this assembly (PayloadCollection) to parse out the version information, so it'll need to be internal